### ISSUE TITLE
NextObserver<T> is not assignable to (value: T) => void

### ISSUE THREAD

**Issue by benlesh** [ReactiveX/rxjs#4150]
## Bug Report

**Current Behavior**
Currently, if you pass `observer, undefined, undefined` to `subscribe` you get a build error.

**Reproduction**
- REPL or Repo link:
(you can use https://stackblitz.com/ to create one to attach here)

```ts
import { Observable, PartialObserver } from 'rxjs';


const o = new Observable();

function foo<T>(
  next?: PartialObserver<T>|((value: T) => void),
  error?: (err: any) => void,
  complete?: () => void,
) {
  o.subscribe(next, error, complete); // ERROR HERE
}
```

**Expected behavior**
No build error.

**Environment**
- RxJS version: 6.3.x

**Possible Solution**
Update typings for `Subscribable` to support passing undefined for the other two arguments.

**Comment by benlesh** [comment]
The work around here is to just cast `next as any` in the `subscribe` call.

```ts
o.subscribe(next as any, error, complete);
```

I'm on the fence about this as a bug. It's cumbersome the way it is now in 6.3... but is it also helpful? Hard to say. We should discuss this.

**Comment by cartant** [comment]
I'm not sure this is a bug. IMO, situations like this can arise anytime you have a function that's implemented using another function that has overload signatures.

But maybe something could be done to address this and also the fact that it should be possible to pass `null` as well as `undefined`. Something like this, perhaps?

```ts
export interface Subscribable<T> {
  subscribe(observer?: PartialObserver<T>, error?: null, complete?: null): Unsubscribable;
  subscribe(next?: (value: T) => void, error?: ((error: any) => void) | null, complete?: (() => void) | null): Unsubscribable;
}
```

That would catch usage like this:

```ts
source.subscribe({ next: value => console.log(value) }, error => console.error(error));
```

but I doubt it would solve this issue's problem.

Also, if support for `null` added, there are a few other places to be updated (e.g. the `do` and `tap` operators).

**Comment by benlesh** [comment]
Well, basically these are probably all okay:

```
o.subscribe(observer);
o.subscribe(fn?, fn?, fn?);
o.subscribe(observer, ...u: undefined[]);
```

and this is likely a problem:

```
o.subscribe(observer, fn)
o.subscribe(observer, null|undefined, fn)
o.subscribe(observer, fn, fn)
```


Realistically, long-term, I'd like to guide people to only use:

```ts
subscribe(observer: PartialObserver<T>): Subscription;

// or

subscribe(nextHandler: (value: T) => void): Subscription
```

... because everything else is simply less readable for no real gain.

**Comment by rkirov** [comment]
The underlying issue is that you are mixing 'union types' and function overloads, and those are not the same thing. These two interfaces behave very differently:

```typescript
interface I1 {
  foo(a: A | B): void;
}

interface I2 {
  foo(a: A): void;
  foo(b: B): void;
}
```

For example:

```typescript
function f(a: A|B, i1: I1, i2: I2) {
  i1.foo(a);  // ok
  i2.foo(a);  // is a type error because neither overload match. One wants A, the other B, but we have A | B.
}
```

In your case `Observable` uses overloads, while `Subscribable` uses union types. You should consistently use one or the other, but I am not sure which one.

For unions, the case is that it is easier to write, as in Ben's example. For overloads, I can see how it is easier for maintainers to separate different number of args and comments.

P.S. I expect the original error go away if you swap with `Subscribable` which is written in the union style, but stackblitz still error. Any ideas why?

**Comment by cartant** [comment]
@rkirov Regarding the history of the signatures, I doubt there is a reason for their being different.

Since I started working with RxJS, `Observable.subscribe` had overload signatures. Originally, it had an empty signature overload, but that caused problems with the typings. [This PR](https://github.com/ReactiveX/rxjs/pull/3053/files) removed it, but still left `Observable.subscribe` with overloads instead of a union.

More recently, [this PR](https://github.com/ReactiveX/rxjs/pull/4050) was merged a while ago to bring the `Subscribable` signatures into line with the `Observable` signatures. However, I'm not sure that PR is included in a published version of RxJS.
